File: Mandelbrot.py
from PIL import Image
from numba import njit


width, height = 3840, 2400

#Used to find position of points in relation to center of image
xMid = width // 2 + width // 4
yMid = height // 2
#Scales the image so the axis are lower to make the set visible
xDivFactor = width / 3
yDivFactor = height / 2

im = Image.new("RGB", (width, height))
pix = im.load()

#Given the coordinates of a point, iterate z**2 + c to see if it goes to infinity
#Fractal is on complex plane so treat y axis as imaginary numbers
#Use JIT compiler to accelerate calculations by about 700%
@njit
def calcMandel(x,y, maxIts):
    z = 0
    c = complex(x/xDivFactor, y/yDivFactor)
    for k in range(maxIts):
        z = z**2 + c
        if abs(z) > 2:
            #Nice cyan colour, intensity depends on how quickly z**2 + c increases past 2
            return ((0, k*10, k*20))
    #If it stays less than 0, it is inside the bulb so remain black
    return ((0,0,0))

def mandelbrot():
    #For each pixel in the image, find if it is part of the set and colour the pixel accordingly
    maxIts = 50
    for i in range(width):
        x = i - xMid
        for j in range(height):
            pix[i,j] = calcMandel(x,j - yMid, maxIts)
    #Write the pixels to a PNG in the same directory as the python file, called Mandelbrot
    im.save("Mandelbrot.png", "PNG")

if __name__ == '__main__':
    mandelbrot()

File: mandelbrot.cpp
#include <complex>
#include <fstream>

int calcMandel(int x, int y, int maxIts, double xDivFactor, double yDivFactor){
    std::complex<double> z = (0.0, 0.0);
    double newX = x / xDivFactor;
    double newY = y / yDivFactor;
    std::complex<double> c;
    //For some reason, doing c = (newX, newY) always gave (-1, 0) so used this instead
    c.real(newX);
    c.imag(newY);
    for (int k = 0; k < maxIts; k++){
        z = pow(z, 2) + c;
        if (abs(z) > 2){
            return k;
        }
    }
    return 0;
}

int main(){
    int width = 3840;
    int height = 2400;
    int xMid = width / 2 + width / 4;
    int yMid = height / 2;
    double xDivFactor = width / 3;
    double yDivFactor = height / 2;
    int maxIts = 25;
    std::ofstream f;
    f.open("mandelbrotVals.txt");
    f << width << "," << height << ",";
    for (int j = 0; j < height; j++){
        for (int i = 0; i < width; i++){
            int x = i - xMid;
            f << calcMandel(x, j - yMid, maxIts, xDivFactor, yDivFactor) << ",";;
        }
    }
    f.close();
}